File: reference/FlaMEVB/modIO.vb
Imports ICSharpCode.SharpZipLib
Imports System.Globalization

Public Module modIO

    Public Function ZipMakeEntry(ZipOutputStream As Zip.ZipOutputStream, Path As String, Result As clsResult) As Zip.ZipEntry

        Try
            Dim NewZipEntry As New Zip.ZipEntry(Path)
            NewZipEntry.DateTime = Now
            ZipOutputStream.PutNextEntry(NewZipEntry)
            Return NewZipEntry
        Catch ex As Exception
            Result.ProblemAdd("Zip entry " & ControlChars.Quote & Path & ControlChars.Quote & " failed: " & ex.Message)
            Return Nothing
        End Try
    End Function

    Public Function InvariantToString_bool(Value As Boolean) As String

        Return Value.ToString(CultureInfo.InvariantCulture)
    End Function

    Public Function InvariantToString_byte(Value As Byte) As String

        Return Value.ToString(CultureInfo.InvariantCulture)
    End Function

    Public Function InvariantToString_short(Value As Short) As String

        Return Value.ToString(CultureInfo.InvariantCulture)
    End Function

    Public Function InvariantToString_int(Value As Integer) As String

        Return Value.ToString(CultureInfo.InvariantCulture)
    End Function

    Public Function InvariantToString_uint(Value As UInteger) As String

        Return Value.ToString(CultureInfo.InvariantCulture)
    End Function

    Public Function InvariantToString_sng(Value As Single) As String

        Return Value.ToString(CultureInfo.InvariantCulture)
    End Function

    Public Function InvariantToString_dbl(Value As Double) As String

        Return Value.ToString(CultureInfo.InvariantCulture)
    End Function

    Public Function InvariantParse_bool(Text As String, ByRef Result As Boolean) As Boolean

        Return Boolean.TryParse(Text, Result)
    End Function

    Public Function InvariantParse_byte(Text As String, ByRef Result As Byte) As Boolean

        Return Byte.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, Result)
    End Function

    Public Function InvariantParse_short(Text As String, ByRef Result As Short) As Boolean

        Return Short.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, Result)
    End Function

    Public Function InvariantParse_ushort(Text As String, ByRef Result As UShort) As Boolean

        Return UShort.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, Result)
    End Function

    Public Function InvariantParse_int(Text As String, ByRef Result As Integer) As Boolean

        Return Integer.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, Result)
    End Function

    Public Function InvariantParse_uint(Text As String, ByRef Result As UInteger) As Boolean

        Return UInteger.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, Result)
    End Function

    Public Function InvariantParse_sng(Text As String, ByRef Result As Single) As Boolean

        Return Single.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, Result)
    End Function

    Public Function InvariantParse_dbl(Text As String, ByRef Result As Double) As Boolean

        Return Double.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, Result)
    End Function

    Public Function ReadOldText(File As IO.BinaryReader) As String
        Dim Result As String = ""
        Dim A As Integer
        Dim Length As Integer = CInt(File.ReadUInt32)

        For A = 0 To Length - 1
            Result &= Chr(File.ReadByte)
        Next
        Return Result
    End Function

    Public Function ReadOldTextOfLength(File As IO.BinaryReader, Length As Integer) As String
        Dim Result As String = ""
        Dim A As Integer

        For A = 0 To Length - 1
            Result &= Chr(File.ReadByte)
        Next
        Return Result
    End Function

    Public Sub WriteText(File As IO.BinaryWriter, WriteLength As Boolean, Text As String)

        If WriteLength Then
            File.Write(CUInt(Text.Length))
        End If
        Dim A As Integer
        For A = 0 To Text.Length - 1
            File.Write(CByte(Asc(Text(A))))
        Next
    End Sub

    Public Sub WriteTextOfLength(File As IO.BinaryWriter, Length As Integer, Text As String)
        Dim A As Integer

        For A = 0 To Math.Min(Text.Length, Length) - 1
            File.Write(CByte(Asc(Text(A))))
        Next
        For A = Text.Length To Length - 1
            File.Write(CByte(0))
        Next
    End Sub

    Public Function WriteMemoryToNewFile(Memory As IO.MemoryStream, Path As String) As clsResult
        Dim ReturnResult As New clsResult("Writing to " & ControlChars.Quote & Path & ControlChars.Quote)

        Dim NewFile As IO.FileStream
        Try
            NewFile = New IO.FileStream(Path, IO.FileMode.CreateNew)
        Catch ex As Exception
            ReturnResult.ProblemAdd(ex.Message)
            Return ReturnResult
        End Try
        Try
            Memory.WriteTo(NewFile)
        Catch ex As Exception
            ReturnResult.ProblemAdd(ex.Message)
            Return ReturnResult
        End Try

        Memory.Close()
        NewFile.Close()

        Return ReturnResult
    End Function

    Public Function WriteMemoryToZipEntryAndFlush(Memory As IO.MemoryStream, Stream As Zip.ZipOutputStream) As clsResult
        Dim ReturnResult As New clsResult("Writing to zip stream")

        Try
            Memory.WriteTo(Stream)
            Memory.Flush()
            Stream.Flush()
            Stream.CloseEntry()
        Catch ex As Exception
            ReturnResult.ProblemAdd(ex.Message)
            Return ReturnResult
        End Try

        Return ReturnResult
    End Function

    Public Function TryOpenFileStream(Path As String, ByRef Output As IO.FileStream) As sResult
        Dim ReturnResult As sResult
        ReturnResult.Success = False
        ReturnResult.Problem = ""

        Try
            Output = New IO.FileStream(Path, IO.FileMode.Open)
        Catch ex As Exception
            Output = Nothing
            ReturnResult.Problem = ex.Message
            Return ReturnResult
        End Try

        ReturnResult.Success = True
        Return ReturnResult
    End Function

    Public Function WZAngleFromINIText(Text As String, ByRef Result As sWZAngle) As Boolean
        Dim VectorText As New clsSplitCommaText(Text)
        Dim WZAngle As sWZAngle

        If VectorText.PartCount <> 3 Then
            Return False
        End If

        If Not InvariantParse_ushort(VectorText.Parts(0), WZAngle.Direction) Then
            Dim ErrorValue As Integer
            If Not InvariantParse_int(VectorText.Parts(0), ErrorValue) Then
                Return False
            End If
            Dim Remainder As Integer
            Dim Multiplier As Integer = Math.DivRem(ErrorValue, INIRotationMax, Remainder)
            Try
                If Remainder < 0 Then
                    WZAngle.Direction = CUShort(Remainder + INIRotationMax)
                Else
                    WZAngle.Direction = CUShort(Remainder)
                End If
            Catch ex As Exception
                Return False
            End Try
            Return True
        End If
        If Not InvariantParse_ushort(VectorText.Parts(1), WZAngle.Pitch) Then
            Return False
        End If
        If Not InvariantParse_ushort(VectorText.Parts(2), WZAngle.Roll) Then
            Return False
        End If
        Result = WZAngle
        Return True
    End Function

    Public Function HealthFromINIText(Text As String, ByRef Result As Integer) As Boolean
        Dim A As Integer
        Dim Health As Integer

        A = Text.IndexOf("%"c)
        If A < 0 Then
            Return False
        End If
        Text = Text.Replace("%", "")
        If Not InvariantParse_int(Text, Health) Then
            Return False
        End If
        If Health < 0 Or Health > 100 Then
            Return False
        End If
        Result = Health
        Return True
    End Function

    Public Function WorldPosFromINIText(Text As String, ByRef Result As clsWorldPos) As Boolean
        Dim VectorText As New clsSplitCommaText(Text)
        Dim A As Integer
        Dim B As Integer

        If VectorText.PartCount <> 3 Then
            Return False
        End If
        Dim Positions(2) As Integer
        For A = 0 To 2
            If InvariantParse_int(VectorText.Parts(A), B) Then
                Positions(A) = B
            Else
                Return False
            End If
        Next
        Result = New clsWorldPos(New sWorldPos(New sXY_int(Positions(0), Positions(1)), Positions(2)))
        Return True
    End Function

    Public Function FindZipEntryFromPath(Path As String, ZipPathToFind As String) As clsZipStreamEntry
        Dim ZipStream As Zip.ZipInputStream
        Dim ZipEntry As Zip.ZipEntry
        Dim FindPath As String = ZipPathToFind.ToLower.Replace("\"c, "/"c)
        Dim ZipPath As String

        ZipStream = New Zip.ZipInputStream(IO.File.OpenRead(Path))
        Do
            Try
                ZipEntry = ZipStream.GetNextEntry
            Catch ex As Exception
                Exit Do
            End Try
            If ZipEntry Is Nothing Then
                Exit Do
            End If

            ZipPath = ZipEntry.Name.ToLower.Replace("\"c, "/"c)
            If ZipPath = FindPath Then
                Dim Result As New clsZipStreamEntry
                Result.Stream = ZipStream
                Result.Entry = ZipEntry
                Return Result
            End If
        Loop
        ZipStream.Close()

        Return Nothing
    End Function

    Public Function BytesToLinesRemoveComments(reader As IO.BinaryReader) As SimpleList(Of String)
        Dim CurrentChar As Char
        Dim CurrentCharExists As Boolean
        Dim InLineComment As Boolean
        Dim InCommentBlock As Boolean
        Dim PrevChar As Char
        Dim PrevCharExists As Boolean
        Dim Line As String = ""
        Dim Result As New SimpleList(Of String)

        Do
MonoContinueDo:
            PrevChar = CurrentChar
            PrevCharExists = CurrentCharExists
            Try
                CurrentChar = reader.ReadChar
                CurrentCharExists = True
            Catch ex As Exception
                CurrentCharExists = False
            End Try
            If CurrentCharExists Then
                Select Case CurrentChar
                    Case ControlChars.Cr, ControlChars.Lf
                        InLineComment = False
                        If PrevCharExists Then
                            Line &= PrevChar
                        End If
                        CurrentCharExists = False

                        If Line.Length > 0 Then
                            Result.Add(Line)
                            Line = ""
                        End If

                        GoTo MonoContinueDo
                    Case "*"c
                        If PrevCharExists And PrevChar = "/"c Then
                            InCommentBlock = True
                            CurrentCharExists = False
                            GoTo MonoContinueDo
                        End If
                    Case "/"c
                        If PrevCharExists Then
                            If PrevChar = "/"c Then
                                InLineComment = True
                                CurrentCharExists = False
                                GoTo MonoContinueDo
                            ElseIf PrevChar = "*"c Then
                                InCommentBlock = False
                                CurrentCharExists = False
                                GoTo MonoContinueDo
                            End If
                        End If
                End Select
            Else
                InLineComment = False
                If PrevCharExists Then
                    Line &= PrevChar
                End If
                If Line.Length > 0 Then
                    Result.Add(Line)
                    Line = ""
                End If

                Exit Do
            End If
            If PrevCharExists Then
                If Not (InCommentBlock Or InLineComment) Then
                    Line &= PrevChar
                End If
            End If
        Loop

        Return Result
    End Function
End Module

Public Class clsPositionFromText

    Public Pos As sXY_int

    Public Function Translate(Text As String) As Boolean
        Dim A As Integer
        Dim Positions As New clsSplitCommaText(Text)

        If Positions.PartCount < 2 Then
            Return False
        End If
        If InvariantParse_int(Positions.Parts(0), A) Then
            Pos.X = A
        Else
            Return False
        End If
        If InvariantParse_int(Positions.Parts(1), A) Then
            Pos.Y = A
        Else
            Return False
        End If
        Return True
    End Function
End Class

Public Class clsZipStreamEntry
    Public Stream As Zip.ZipInputStream
    Public Entry As Zip.ZipEntry
End Class

Public Class clsSplitCommaText

    Public Parts() As String
    Public PartCount As Integer

    Public Sub New(Text As String)
        Dim A As Integer

        Parts = Text.Split(","c)
        PartCount = Parts.GetUpperBound(0) + 1
        For A = 0 To PartCount - 1
            Parts(A) = Parts(A).Trim()
        Next
    End Sub
End Class

File: source/SharpFlame/modIO.cs
using System;
using System.Globalization;
using System.IO;
using ICSharpCode.SharpZipLib.Zip;
using Microsoft.VisualBasic;
using SharpFlame.Collections;

namespace SharpFlame
{
    public sealed class modIO
    {
        public static ZipEntry ZipMakeEntry(ZipOutputStream ZipOutputStream, string Path, clsResult Result)
        {
            try
            {
                ZipEntry NewZipEntry = new ZipEntry(Path);
                NewZipEntry.DateTime = DateTime.Now;
                ZipOutputStream.PutNextEntry(NewZipEntry);
                return NewZipEntry;
            }
            catch ( Exception ex )
            {
                Result.ProblemAdd("Zip entry " + Convert.ToString(ControlChars.Quote) + Path + Convert.ToString(ControlChars.Quote) + " failed: " +
                                  ex.Message);
                return null;
            }
        }

        public static string InvariantToString_bool(bool Value)
        {
            return Value.ToString(CultureInfo.InvariantCulture);
        }

        public static string InvariantToString_byte(byte Value)
        {
            return Value.ToString(CultureInfo.InvariantCulture);
        }

        public static string InvariantToString_short(short Value)
        {
            return Value.ToString(CultureInfo.InvariantCulture);
        }

        public static string InvariantToString_int(int Value)
        {
            return Value.ToString(CultureInfo.InvariantCulture);
        }

        public static string InvariantToString_uint(UInt32 Value)
        {
            return Value.ToString(CultureInfo.InvariantCulture);
        }

        public static string InvariantToString_sng(float Value)
        {
            return Value.ToString(CultureInfo.InvariantCulture);
        }

        public static string InvariantToString_dbl(double Value)
        {
            return Value.ToString(CultureInfo.InvariantCulture);
        }

        public static bool InvariantParse_bool(string Text, ref bool Result)
        {
            return bool.TryParse(Text, out Result);
        }

        public static bool InvariantParse_byte(string Text, ref byte Result)
        {
            return byte.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, out Result);
        }

        public static bool InvariantParse_short(string Text, ref short Result)
        {
            return short.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, out Result);
        }

        public static bool InvariantParse_ushort(string Text, UInt16 Result)
        {
            return UInt16.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, out Result);
        }

        public static bool InvariantParse_int(string Text, ref int Result)
        {
            return int.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, out Result);
        }

        public static bool InvariantParse_uint(string Text, UInt32 Result)
        {
            return UInt32.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, out Result);
        }

        public static bool InvariantParse_sng(string Text, ref float Result)
        {
            return float.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, out Result);
        }

        public static bool InvariantParse_dbl(string Text, ref double Result)
        {
            return double.TryParse(Text, NumberStyles.Any, CultureInfo.InvariantCulture, out Result);
        }

        public static string ReadOldText(BinaryReader File)
        {
            string Result = "";
            int A = 0;
            int Length = (int)(File.ReadUInt32());

            for ( A = 0; A <= Length - 1; A++ )
            {
                Result += Convert.ToString(Strings.Chr(File.ReadByte()));
            }
            return Result;
        }

        public static string ReadOldTextOfLength(BinaryReader File, int Length)
        {
            string Result = "";
            int A = 0;

            for ( A = 0; A <= Length - 1; A++ )
            {
                Result += Convert.ToString(Strings.Chr(File.ReadByte()));
            }
            return Result;
        }

        public static void WriteText(BinaryWriter File, bool WriteLength, string Text)
        {
            if ( WriteLength )
            {
                File.Write(Convert.ToBoolean((uint)Text.Length));
            }
            int A = 0;
            for ( A = 0; A <= Text.Length - 1; A++ )
            {
                File.Write((byte)(Strings.Asc(Text[A])));
            }
        }

        public static void WriteTextOfLength(BinaryWriter File, int Length, string Text)
        {
            int A = 0;

            for ( A = 0; A <= Math.Min(Text.Length, Length) - 1; A++ )
            {
                File.Write((byte)(Strings.Asc(Text[A])));
            }
            for ( A = Text.Length; A <= Length - 1; A++ )
            {
                File.Write((byte)0);
            }
        }

        public static clsResult WriteMemoryToNewFile(MemoryStream Memory, string Path)
        {
            clsResult ReturnResult = new clsResult("Writing to " + Convert.ToString(ControlChars.Quote) + Path + Convert.ToString(ControlChars.Quote));

            FileStream NewFile = default(FileStream);
            try
            {
                NewFile = new FileStream(Path, FileMode.CreateNew);
            }
            catch ( Exception ex )
            {
                ReturnResult.ProblemAdd(ex.Message);
                return ReturnResult;
            }
            try
            {
                Memory.WriteTo(NewFile);
            }
            catch ( Exception ex )
            {
                ReturnResult.ProblemAdd(ex.Message);
                return ReturnResult;
            }

            Memory.Close();
            NewFile.Close();

            return ReturnResult;
        }

        public static clsResult WriteMemoryToZipEntryAndFlush(MemoryStream Memory, ZipOutputStream Stream)
        {
            clsResult ReturnResult = new clsResult("Writing to zip stream");

            try
            {
                Memory.WriteTo(Stream);
                Memory.Flush();
                Stream.Flush();
                Stream.CloseEntry();
            }
            catch ( Exception ex )
            {
                ReturnResult.ProblemAdd(ex.Message);
                return ReturnResult;
            }

            return ReturnResult;
        }

        public static modProgram.sResult TryOpenFileStream(string Path, ref FileStream Output)
        {
            modProgram.sResult ReturnResult = new modProgram.sResult();
            ReturnResult.Success = false;
            ReturnResult.Problem = "";

            try
            {
                Output = new FileStream(Path, FileMode.Open);
            }
            catch ( Exception ex )
            {
                Output = null;
                ReturnResult.Problem = ex.Message;
                return ReturnResult;
            }

            ReturnResult.Success = true;
            return ReturnResult;
        }

        public static bool WZAngleFromINIText(string Text, ref modProgram.sWZAngle Result)
        {
            clsSplitCommaText VectorText = new clsSplitCommaText(Text);
            modProgram.sWZAngle WZAngle = new modProgram.sWZAngle();

            if ( VectorText.PartCount != 3 )
            {
                return false;
            }

            if ( !InvariantParse_ushort(VectorText.Parts[0], WZAngle.Direction) )
            {
                int ErrorValue = 0;
                if ( !InvariantParse_int(VectorText.Parts[0], ref ErrorValue) )
                {
                    return false;
                }
                int Remainder = 0;
                int Multiplier = Math.DivRem(ErrorValue, modProgram.INIRotationMax, out Remainder);
                try
                {
                    if ( Remainder < 0 )
                    {
                        WZAngle.Direction = (ushort)(Remainder + modProgram.INIRotationMax);
                    }
                    else
                    {
                        WZAngle.Direction = (ushort)Remainder;
                    }
                }
                catch ( Exception )
                {
                    return false;
                }
                return true;
            }
            if ( !InvariantParse_ushort(VectorText.Parts[1], WZAngle.Pitch) )
            {
                return false;
            }
            if ( !InvariantParse_ushort(VectorText.Parts[2], WZAngle.Roll) )
            {
                return false;
            }
            Result = WZAngle;
            return true;
        }

        public static bool HealthFromINIText(string Text, ref int Result)
        {
            int A = 0;
            int Health = 0;

            A = Text.IndexOf('%');
            if ( A < 0 )
            {
                return false;
            }
            Text = Text.Replace("%", "");
            if ( !InvariantParse_int(Text, ref Health) )
            {
                return false;
            }
            if ( Health < 0 | Health > 100 )
            {
                return false;
            }
            Result = Health;
            return true;
        }

        public static bool WorldPosFromINIText(string Text, ref modProgram.clsWorldPos Result)
        {
            clsSplitCommaText VectorText = new clsSplitCommaText(Text);
            int A = 0;
            int B = 0;

            if ( VectorText.PartCount != 3 )
            {
                return false;
            }
            int[] Positions = new int[3];
            for ( A = 0; A <= 2; A++ )
            {
                if ( InvariantParse_int(VectorText.Parts[A], ref B) )
                {
                    Positions[A] = B;
                }
                else
                {
                    return false;
                }
            }
            Result = new modProgram.clsWorldPos(new modProgram.sWorldPos(new modMath.sXY_int(Positions[0], Positions[1]), Positions[2]));
            return true;
        }

        public static clsZipStreamEntry FindZipEntryFromPath(string Path, string ZipPathToFind)
        {
            ZipInputStream ZipStream = default(ZipInputStream);
            ZipEntry ZipEntry = default(ZipEntry);
            string FindPath = ZipPathToFind.ToLower().Replace('\\', '/');
            string ZipPath;

            ZipStream = new ZipInputStream(File.OpenRead(Path));
            do
            {
                try
                {
                    ZipEntry = ZipStream.GetNextEntry();
                }
                catch ( Exception )
                {
                    goto endOfDoLoop;
                }
                if ( ZipEntry == null )
                {
                    break;
                }

                ZipPath = ZipEntry.Name.ToLower().Replace('\\', '/');
                if ( ZipPath == FindPath )
                {
                    clsZipStreamEntry Result = new clsZipStreamEntry();
                    Result.Stream = ZipStream;
                    Result.Entry = ZipEntry;
                    return Result;
                }
            } while ( true );
            endOfDoLoop:
            ZipStream.Close();

            return null;
        }

        public static SimpleList<string> BytesToLinesRemoveComments(BinaryReader reader)
        {
            char CurrentChar = (char)0;
            bool CurrentCharExists = default(bool);
            bool InLineComment = default(bool);
            bool InCommentBlock = default(bool);
            char PrevChar = (char)0;
            bool PrevCharExists = default(bool);
            string Line = "";
            SimpleList<string> Result = new SimpleList<string>();

            do
            {
                MonoContinueDo:
                PrevChar = CurrentChar;
                PrevCharExists = CurrentCharExists;
                try
                {
                    CurrentChar = reader.ReadChar();
                    CurrentCharExists = true;
                }
                catch ( Exception )
                {
                    CurrentCharExists = false;
                }
                if ( CurrentCharExists )
                {
                    switch ( CurrentChar )
                    {
                        case ControlChars.Cr:
                        case ControlChars.Lf:
                            InLineComment = false;
                            if ( PrevCharExists )
                            {
                                Line += PrevChar.ToString();
                            }
                            CurrentCharExists = false;

                            if ( Line.Length > 0 )
                            {
                                Result.Add(Line);
                                Line = "";
                            }

                            goto MonoContinueDo;
                        case '*':
                            if ( PrevCharExists && PrevChar == '/' )
                            {
                                InCommentBlock = true;
                                CurrentCharExists = false;
                                goto MonoContinueDo;
                            }
                            break;
                        case '/':
                            if ( PrevCharExists )
                            {
                                if ( PrevChar == '/' )
                                {
                                    InLineComment = true;
                                    CurrentCharExists = false;
                                    goto MonoContinueDo;
                                }
                                else if ( PrevChar == '*' )
                                {
                                    InCommentBlock = false;
                                    CurrentCharExists = false;
                                    goto MonoContinueDo;
                                }
                            }
                            break;
                    }
                }
                else
                {
                    InLineComment = false;
                    if ( PrevCharExists )
                    {
                        Line += PrevChar.ToString();
                    }
                    if ( Line.Length > 0 )
                    {
                        Result.Add(Line);
                        Line = "";
                    }

                    break;
                }
                if ( PrevCharExists )
                {
                    if ( !(InCommentBlock || InLineComment) )
                    {
                        Line += PrevChar.ToString();
                    }
                }
            } while ( true );

            return Result;
        }
    }

    public class clsPositionFromText
    {
        public modMath.sXY_int Pos;

        public bool Translate(string Text)
        {
            int A = 0;
            clsSplitCommaText Positions = new clsSplitCommaText(Text);

            if ( Positions.PartCount < 2 )
            {
                return false;
            }
            if ( modIO.InvariantParse_int(Positions.Parts[0], ref A) )
            {
                Pos.X = A;
            }
            else
            {
                return false;
            }
            if ( modIO.InvariantParse_int(Positions.Parts[1], ref A) )
            {
                Pos.Y = A;
            }
            else
            {
                return false;
            }
            return true;
        }
    }

    public class clsZipStreamEntry
    {
        public ZipInputStream Stream;
        public ZipEntry Entry;
    }

    public class clsSplitCommaText
    {
        public string[] Parts;
        public int PartCount;

        public clsSplitCommaText(string Text)
        {
            int A = 0;

            Parts = Text.Split(',');
            PartCount = Parts.GetUpperBound(0) + 1;
            for ( A = 0; A <= PartCount - 1; A++ )
            {
                Parts[A] = Convert.ToString(Parts[A].Trim());
            }
        }
    }
}